### PR TITLE
remove unions in cali_tc_ctx/state structure

### PR DESCRIPTION
**Pull request by mazdakn**
## Description

This PR corrects the type punning in bpf code by removing the unions in the `cai_tc_ctx` structure.

## Todos
- [x] Unit tests (full coverage)

## Release Note

```release-note
None required
```

### CODE REVIEW

File: bpf-gpl/parsing.h [hunk 1]
```diff
@@ -22,6 +22,11 @@
 #define PARSING_ERROR -1
 #define PARSING_ALLOW_WITHOUT_ENFORCING_POLICY -2
 
+#define ETHHDR(x) ((struct ethhdr *)x->data_start)
```
Comment (tomastigera) [inline bpf-gpl/parsing.h:25]: nit: these only work on tc context, what if we prefix them with TC_ ? What if they were functions that return the typecasted type and take typed cali_tc_ctx input? Would it work?

Why not to define them in types.h where the cali_tc_ctx is defined?

**Review by tomastigera — COMMENTED** [review]
looks OK to me

File: bpf-gpl/parsing.h (hunk 1 above)
Comment (mazdakn) [reply bpf-gpl/parsing.h:25]: Just noticed that in types.h only the structures are defined, and the rest of functions are either in skb.h or parsing.h. It looks like those 3 files has a lot in common. The functions are little scattered, for example `ip_dec_ttl()` is in bpf.h. Maybe we should sort them out?

Comment (mazdakn) [reply bpf-gpl/parsing.h:25]: > these only work on tc context, what if we prefix them with TC_ ? What if they were functions that return the typecasted type and take typed cali_tc_ctx input? Would it work?
That would definitely work. Honestly between macros and functions, I prefer functions:-D I'll switch to function.

Comment (mazdakn) [reply bpf-gpl/parsing.h]: @tomastigera I think you right. it is better to move these to types.h.

**Review by tomastigera — APPROVED** [review]
LGTM thanks

**Pull request merged**